File: arch/x86/cgc.go
package x86

import (
	"crypto/rand"
	uc "github.com/lunixbochs/unicorn"
	"syscall"

	"../../models"
)

func writeAddr(u models.Usercorn, addr, val uint64) {
	var buf [4]byte
	u.PackAddr(buf[:], val)
	u.MemWrite(addr, buf[:])
}

func CgcSyscall(u models.Usercorn) {
	// TODO: handle errors or something
	args, _ := u.ReadRegs(LinuxRegs)
	eax, _ := u.RegRead(uc.UC_X86_REG_EAX)
	var ret uint64
	switch eax {
	case 1: // _terminate
		syscall.Exit(int(args[0]))
	case 2: // transmit
		mem, _ := u.MemRead(args[1], args[2])
		n, _ := syscall.Write(int(args[0]), mem)
		writeAddr(u, args[3], uint64(n))
	case 3: // receive
		tmp := make([]byte, args[2])
		n, _ := syscall.Read(int(args[0]), tmp)
		u.MemWrite(args[1], tmp[:n])
		writeAddr(u, args[3], uint64(n))
	case 5: // allocate
		addr, _ := u.Mmap(0, args[0])
		// args[1] == is executable
		writeAddr(u, args[2], addr)
	case 7: // random
		tmp := make([]byte, args[1])
		rand.Read(tmp)
		u.MemWrite(args[0], tmp)
		writeAddr(u, args[2], args[1])
	}
	u.RegWrite(uc.UC_X86_REG_EAX, ret)
}

func CgcInterrupt(u models.Usercorn, intno uint32) {
	if intno == 0x80 {
		CgcSyscall(u)
	}
}

func init() {
	Arch.RegisterOS(&models.OS{Name: "cgc", Syscall: CgcSyscall, Interrupt: CgcInterrupt})
}

File: py/usercorn/arch/x86/cgc.py
from unicorn.x86_const import *
import os
import sys

def syscall(cls):
    regs = [UC_X86_REG_EAX, UC_X86_REG_EBX, UC_X86_REG_ECX, UC_X86_REG_EDX, UC_X86_REG_ESI, UC_X86_REG_EDI, UC_X86_REG_EBP]
    num, a1, a2, a3, a4, a5, a6 = [cls.reg_read(r) for r in regs]
    ret = 0
    # _terminate
    if num == 1:
        sys.exit(a1)
    # transmit
    elif num == 2:
        n = os.write(a1, cls.mem_read(a2, a3))
        cls.mem_write(a4, cls.pack_addr(n))
    # receive
    elif num == 3: 
        tmp = os.read(a1, a3)
        cls.mem_write(a2, tmp)
        cls.mem_write(a4, cls.pack_addr(len(tmp)))
    # fdwait
    elif num == 4:
        pass
    # allocate
    elif num == 5:
        addr = cls.mmap(a1)
        cls.mem_write(a3, cls.pack_addr(addr))
    # deallocate
    elif num == 6:
        pass
    # random
    elif num == 7: 
        rnd = os.urandom(a2)
        cls.mem_write(a1, rnd)
        cls.mem_write(a3, cls.pack_addr(a2))
    else:
        print 'Unsupported syscall:', num
        sys.exit(1)
    cls.reg_write(UC_X86_REG_EAX, ret)

def interrupt(cls, intno):
    if intno == 0x80:
        syscall(cls)
    else:
        raise NotImplementedError('unhandled interrupt %d' % intno)
